### PR TITLE
Allow etcd snapshot restoration on a fresh set of etcd nodes

### PR DESCRIPTION
**Pull request by Oats87**
## Issue: 
https://github.com/rancher/rancher/issues/41080

Also:
https://github.com/rancher/rancher/issues/41174
https://github.com/rancher/rancher/issues/40994

Partially:
https://github.com/rancher/rancher/issues/41024
 
## Problem
Today, it is impossible to restore a K3s/RKE2 cluster from etcd snapshot if you lose the underlying machines.
 
## Solution
This PR adds quite a bit of functionality to the planner. It introduces a few new conditions (namely `Bootstrapped` and `Stable`) that is used to indicate whether the cluster has been bootstrapped and stable, and uses this condition to block reconciliation of a cluster if there are no "alive" etcd machines within the cluster. It adds the ability to skip safe removal of an etcd node for emergency purposes, and uses this functionality to allow for complete removal of the etcd plane if the cluster is no longer active (for the purposes of recovery).

The `Stable` condition is `True` when reconciliation is complete (this means, plans were delivered to all nodes and all nodes pass probes), and set back to False when the cluster is not sane (not at least 1 non-deleting etcd/controlplane/worker node). 

The `Boostrapped` condition is used to indicate whether the cluster was bootstrapped. It is set to True once the bootstrap tier is reconciled, and set back to False if the cluster is restored from etcd snapshot (and then re-set to True once snapshot restoration is successful).

It now provides feedback via the UI that indicates the cluster needs to be restored from etcd snapshot when the conditions are right.

It changes the source of the "Ready" condition on the UI object (either v3.Cluster.status condition `Ready` if the cluster is bootstrapped, or v1.RKEControlPlane.status condition `Ready` based on the `Stable` condition of the `rkecontrolplane`. 

It now properly uninitializes/unready the rkecontrolplane and the CAPI cluster when the cluster is being restored from snapshot.

It moves etcd node safe removal from happening on the inframachine objects to happening on the rkebootstrap objects, which is a requirement to ensure that there is controlplane/bootstrap provider compatibility with "other" infrastructure providers.

It adds an additional argument during etcd snapshot restoration that specifies the advertise-client-urls for etcd, which is a permanent workaround for: https://github.com/rancher/rke2/issues/4052

It removes the RKE2 server manifests that match `rke2-*.yaml` to help solve 41174.

It changes etcd restoration reconciliation logic to using the same exact logic as the planner when it delivers plans, notably while ignoring concurrency/drain.

It adds helper functions to the planEntry file that can be used to reduce the number of roleFilters we have.

It checks the S3 endpoint CA argument to determine if it is a filepath or the actual certificate, and operates according to the type. 

### Manual Testing
I manually tested this and was able to restore from S3 snapshots on a new etcd node.
Due to K3s/RKE2 bugs, I was unable to "easily" test restoration from local snapshot to a new node but there is a workaround for this.

### Automated Testing
I added a unit test to test the refactored condition reconciliation function.

Additional automated testing will be added to test etcd restoration in the existing "provisioning-tests" in a future PR.

### CODE REVIEW

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go [hunk 1]
```diff
@@ -373,3 +379,85 @@ func getLabelsAndAnnotationsForPlanSecret(bootstrap *rkev1.RKEBootstrap, machine
...
+
+	if machine.Status.NodeRef == nil {
+		logrus.Infof("[rkebootstrap] No associated node found for machine %s in cluster %s, proceeding with removal", key, clusterName)
+		return bootstrap, nil
+	}
+
+	kcSecret, err := h.secretCache.Get(bootstrap.Namespace, secret.Name(clusterName, secret.Kubeconfig))
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return bootstrap, fmt.Errorf("error retrieving kubeconfig secret %s/%s: %v", bootstrap.Namespace, secret.Name(clusterName, secret.Kubeconfig), err)
```
[the remark this answers is not in the document]
Comment (Oats87) [reply pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:444]: this is likely a remnant from some other logic that I had at one point. No reason to add the color, the error message should tell us the namespace/secret that it was trying to delete

File: pkg/controllers/provisioningv2/rke2/provisioningcluster/controller.go [hunk 2]
```diff
@@ -379,73 +395,31 @@ func (h *handler) getRKEControlPlaneForCluster(cluster *rancherv1.Cluster) (*rke
...
-		clusterStatus, err = h.updateClusterProvisioningStatus(cluster, clusterStatus, cp, rke2.Provisioned, rke2.Ready)
-		if err != nil {
-			return clusterStatus, err
-		}
-
-		return clusterStatus, generic.ErrSkip
+// reconcileCondition accepts an object, a condition to reconcile on the object, and the status/reason/message to reconcile.
+// It returns a boolean indicating whether the condition was changed.
+func reconcileCondition(to interface{}, toCondition condition.Cond, from interface{}, fromCondition condition.Cond) bool {
+	if fromCondition.GetStatus(from) != toCondition.GetStatus(to) || fromCondition.GetReason(from) != toCondition.GetMessage(to) || fromCondition.GetMessage(from) != toCondition.GetMessage(to) {
```
[the remark this answers is not in the document]
Comment (Oats87) [reply pkg/controllers/provisioningv2/rke2/provisioningcluster/controller.go:416]: this function is designed to copy the status/reason/message from the "from" argument to the "to" argument.

For example, I would expect 
if I have Object A with a condition "MyCondition" with message "wow!" and Object B with no conditions,  I could do 

```
reconcileCondition(&b, condition.Cond("SomeOtherCondition"), &a, condition.Cond("MyCondition"))
```

and my result would be that Object B has `SomeOtherCondition.message = "wow!"`.

I wrote a simple example to demontrate this: 
```
package main

import (
	"fmt"

	"github.com/rancher/wrangler/pkg/condition"
	"github.com/rancher/wrangler/pkg/genericcondition"
)

type myObjectStatus struct {
	Conditions []genericcondition.GenericCondition
}

type myObject struct {
	Status myObjectStatus
}

// reconcileCondition accepts an object, a condition to reconcile on the object, and the status/reason/message to reconcile.
// It returns a boolean indicating whether the condition was changed.
func reconcileCondition(to interface{}, toCondition condition.Cond, from interface{}, fromCondition condition.Cond) bool {
	if fromCondition.GetStatus(from) != toCondition.GetStatus(to) || fromCondition.GetReason(from) != toCondition.GetMessage(to) || fromCondition.GetMessage(from) != toCondition.GetMessage(to) {
		toCondition.SetStatus(to, fromCondition.GetStatus(from))
		toCondition.Reason(to, fromCondition.GetReason(from))
		toCondition.Message(to, fromCondition.GetMessage(from))
		return true
	}
	return false
}

func main() {
	MyCondition := condition.Cond("MyCondition")
	SomeOtherCondition := condition.Cond("SomeOtherCondition")

	var a = myObject{}
	var b = myObject{}

	MyCondition.Message(&a, "wow!")
	fmt.Println(fmt.Sprintf("MyCondition on a before reconcile was:\t\t%s", MyCondition.GetMessage(&a)))
	fmt.Println(fmt.Sprintf("SomeOtherCondition on a before reconcile was:\t%s", SomeOtherCondition.GetMessage(&a)))
	fmt.Println()
	fmt.Println(fmt.Sprintf("MyCondition on b before reconcile was:\t\t%s", MyCondition.GetMessage(&b)))
	fmt.Println(fmt.Sprintf("SomeOtherCondition on b before reconcile was:\t%s", SomeOtherCondition.GetMessage(&b)))
	fmt.Println()
	fmt.Println(fmt.Sprintf("reconcileCondition returned\t%s", reconcileCondition(&b, SomeOtherCondition, &a, MyCondition)))
	fmt.Println()
	fmt.Println(fmt.Sprintf("MyCondition on a after reconcile was:\t\t%s", MyCondition.GetMessage(&a)))
	fmt.Println(fmt.Sprintf("SomeOtherCondition on a after reconcile was:\t%s", SomeOtherCondition.GetMessage(&a)))
	fmt.Println()
	fmt.Println(fmt.Sprintf("MyCondition on b after reconcile was:\t\t%s", MyCondition.GetMessage(&b)))
	fmt.Println(fmt.Sprintf("SomeOtherCondition on b after reconcile was:\t%s", SomeOtherCondition.GetMessage(&b)))
}
```

and the output I get from this is:
```
[root@cannon01 test]# go build .
[root@cannon01 test]# ./test
MyCondition on a before reconcile was:		wow!
SomeOtherCondition on a before reconcile was:

MyCondition on b before reconcile was:
SomeOtherCondition on b before reconcile was:

reconcileCondition returned	%!s(bool=true)

MyCondition on a after reconcile was:		wow!
SomeOtherCondition on a after reconcile was:

MyCondition on b after reconcile was:
SomeOtherCondition on b after reconcile was:	wow!
```

Hope this clarifies.

File: pkg/provisioningv2/rke2/planner/etcdrestore.go [hunk 3]
```diff
@@ -53,16 +58,19 @@ func (p *Planner) runEtcdSnapshotRestorePlan(controlPlane *rkev1.RKEControlPlane
...
-		if isS3 ||
-			(server.Machine.Labels[rke2.MachineIDLabel] != "" && snapshot.Labels[rke2.MachineIDLabel] != "" &&
-				server.Machine.Labels[rke2.MachineIDLabel] == snapshot.Labels[rke2.MachineIDLabel]) {
-			restorePlan, joinedServer, err := p.generateEtcdSnapshotRestorePlan(controlPlane, snapshot, tokensSecret, server, joinServer)
-			if err != nil {
-				return err
-			}
-			return assignAndCheckPlan(p.store, ETCDRestoreMessage, server, restorePlan, joinedServer, 0, 0)
+	if len(servers) != 1 {
+		return fmt.Errorf("more than one init node existed, cannot running etcd snapshot restore")
```
[the remark this answers is not in the document]
Comment (Oats87) [reply pkg/provisioningv2/rke2/planner/etcdrestore.go:62]: great catch!

**Review by jakefhyde — COMMENTED** [review]
I'm not done reviewing but I may as well leave this comment lol

File: pkg/provisioningv2/rke2/planner/planentry.go [hunk 4]
```diff
@@ -129,26 +129,26 @@ func anyRole(entry *planEntry) bool {
...
-func notWindows(entry *planEntry) bool {
-	return entry.Machine.Status.NodeInfo.OperatingSystem != windows
+func windows(entry *planEntry) bool {
+	if entry == nil || entry.Machine == nil {
+		return false
+	}
+	if val, ok := entry.Machine.Labels[rke2.CattleOSLabel]; ok {
+		return val == rke2.WindowsMachineOS
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (Oats87) [reply pkg/provisioningv2/rke2/planner/planentry.go]: good call out.

we can of course always enhance the system-agent install script to deliver the OS it is running on or something...

[the remark this answers is not in the document]
Comment (jakefhyde) [reply pkg/provisioningv2/rke2/planner/planentry.go]: Maybe we could add a field on the bootstrap status and populate that to the plan secret?

File: pkg/provisioningv2/rke2/planner/etcdrestore.go [hunk 5]
```diff
@@ -301,36 +349,58 @@ func (p *Planner) restoreEtcdSnapshot(cp *rkev1.RKEControlPlane, status rkev1.RK
...
+	if snapshot != nil {
+		clusterSpec, err := rke2.ParseSnapshotClusterSpecOrError(snapshot)
+		if err != nil || clusterSpec == nil {
+			logrus.Errorf("[planner] rkecluster %s/%s: error parsing snapshot cluster spec for snapshot %s/%s during etcd restoration: %v", cp.Namespace, cp.Name, snapshot.Namespace, snapshot.Name, err)
+		} else {
+			snapshotK8sVersion, err := semver.NewVersion(clusterSpec.KubernetesVersion)
+			if err != nil {
+				return status, err
+			}
+			if !currentVersion.LessThan(managesystemagent.Kubernetes125) && snapshotK8sVersion.LessThan(managesystemagent.Kubernetes125) {
```
Comment (HarrisonWAffel) [inline pkg/provisioningv2/rke2/planner/etcdrestore.go:367]: makes sense, we'll want to release note this requirement if we haven't already

**Review by HarrisonWAffel — APPROVED** [review]
other than an incomplete function comment this lgtm

**Review by jakefhyde — CHANGES_REQUESTED** [review]
just nits

**Review by jakefhyde — APPROVED** [review]

**Pull request merged**